File: modmenu/modmenu.lua
-- ModMenu UI System
local ModMenu = {}
ModMenu.Version = "1.0.0"
ModMenu.IsOpen = false
ModMenu.Pages = {}
ModMenu.CurrentPage = 1
ModMenu.Elements = {}
ModMenu.Buttons = {}

-- Initialize ModMenu
function ModMenu:Initialize(config)
    self.Config = config or {}
    self.Config.Title = self.Config.Title or "MOD MENU"
    self.Config.Width = self.Config.Width or 400
    self.Config.Height = self.Config.Height or 600
    self.Config.Theme = self.Config.Theme or "dark"
    
    print("[ModMenu] Initialized v" .. self.Version)
    return self
end

-- Create a button element
function ModMenu:CreateButton(id, label, callback, config)
    local button = {
        id = id,
        label = label,
        callback = callback,
        color = config and config.color or "#4CAF50",
        width = config and config.width or "100%",
        height = config and config.height or "40px",
        enabled = true
    }
    
    self.Buttons[id] = button
    return button
end

-- Create a toggle element
function ModMenu:CreateToggle(id, label, default, callback, config)
    local toggle = {
        id = id,
        label = label,
        state = default or false,
        callback = callback,
        color = config and config.color or "#2196F3",
        enabled = true
    }
    
    self.Elements[id] = toggle
    return toggle
end

-- Create a slider element
function ModMenu:CreateSlider(id, label, min, max, default, callback, config)
    local slider = {
        id = id,
        label = label,
        min = min,
        max = max,
        value = default or min,
        callback = callback,
        color = config and config.color or "#FF9800",
        enabled = true
    }
    
    self.Elements[id] = slider
    return slider
end

-- Create a dropdown element
function ModMenu:CreateDropdown(id, label, options, default, callback, config)
    local dropdown = {
        id = id,
        label = label,
        options = options,
        selected = default or options[1],
        callback = callback,
        color = config and config.color or "#9C27B0",
        isOpen = false,
        enabled = true
    }
    
    self.Elements[id] = dropdown
    return dropdown
end

-- Create a text input element
function ModMenu:CreateTextInput(id, label, placeholder, callback, config)
    local input = {
        id = id,
        label = label,
        placeholder = placeholder or "Enter text...",
        value = "",
        callback = callback,
        color = config and config.color or "#607D8B",
        enabled = true
    }
    
    self.Elements[id] = input
    return input
end

-- Create a page
function ModMenu:CreatePage(id, title)
    local page = {
        id = id,
        title = title,
        elements = {}
    }
    
    self.Pages[id] = page
    return page
end

-- Add element to page
function ModMenu:AddElementToPage(pageId, elementId)
    if self.Pages[pageId] then
        table.insert(self.Pages[pageId].elements, elementId)
    end
end

-- Toggle menu visibility
function ModMenu:Toggle()
    self.IsOpen = not self.IsOpen
    print("[ModMenu] Menu is now " .. (self.IsOpen and "OPEN" or "CLOSED"))
    return self.IsOpen
end

-- Execute button callback
function ModMenu:PressButton(buttonId)
    if self.Buttons[buttonId] and self.Buttons[buttonId].enabled then
        if self.Buttons[buttonId].callback then
            self.Buttons[buttonId].callback()
        end
        print("[ModMenu] Button pressed: " .. buttonId)
    end
end

-- Toggle switch state
function ModMenu:ToggleSwitch(toggleId)
    if self.Elements[toggleId] and self.Elements[toggleId].state ~= nil then
        self.Elements[toggleId].state = not self.Elements[toggleId].state
        if self.Elements[toggleId].callback then
            self.Elements[toggleId].callback(self.Elements[toggleId].state)
        end
        print("[ModMenu] Toggle " .. toggleId .. " set to " .. tostring(self.Elements[toggleId].state))
    end
end

-- Set slider value
function ModMenu:SetSlider(sliderId, value)
    if self.Elements[sliderId] and self.Elements[sliderId].max then
        local min = self.Elements[sliderId].min
        local max = self.Elements[sliderId].max
        self.Elements[sliderId].value = math.max(min, math.min(max, value))
        
        if self.Elements[sliderId].callback then
            self.Elements[sliderId].callback(self.Elements[sliderId].value)
        end
        print("[ModMenu] Slider " .. sliderId .. " set to " .. tostring(self.Elements[sliderId].value))
    end
end

-- Set dropdown value
function ModMenu:SetDropdown(dropdownId, value)
    if self.Elements[dropdownId] then
        self.Elements[dropdownId].selected = value
        if self.Elements[dropdownId].callback then
            self.Elements[dropdownId].callback(value)
        end
        print("[ModMenu] Dropdown " .. dropdownId .. " set to " .. value)
    end
end

-- Set text input value
function ModMenu:SetTextInput(inputId, value)
    if self.Elements[inputId] and self.Elements[inputId].value ~= nil then
        self.Elements[inputId].value = value
        if self.Elements[inputId].callback then
            self.Elements[inputId].callback(value)
        end
        print("[ModMenu] Text input " .. inputId .. " set to " .. value)
    end
end

-- Get element value
function ModMenu:GetElementValue(elementId)
    if self.Elements[elementId] then
        if self.Elements[elementId].state ~= nil then
            return self.Elements[elementId].state
        elseif self.Elements[elementId].value then
            return self.Elements[elementId].value
        elseif self.Elements[elementId].selected then
            return self.Elements[elementId].selected
        end
    end
    return nil
end

-- Disable/Enable element
function ModMenu:SetElementEnabled(elementId, enabled)
    if self.Elements[elementId] then
        self.Elements[elementId].enabled = enabled
    elseif self.Buttons[elementId] then
        self.Buttons[elementId].enabled = enabled
    end
end

-- Show notification
function ModMenu:Notify(title, message, duration, color)
    local notification = {
        title = title,
        message = message,
        duration = duration or 3,
        color = color or "#4CAF50",
        timestamp = os.time()
    }
    print("[ModMenu Notification] " .. title .. ": " .. message)
    return notification
end

-- Print menu status
function ModMenu:PrintStatus()
    print("\n========== ModMenu Status ==========")
    print("Title: " .. self.Config.Title)
    print("Version: " .. self.Version)
    print("Is Open: " .. tostring(self.IsOpen))
    print("Theme: " .. self.Config.Theme)
    print("Total Pages: " .. tostring(#self.Pages))
    print("Total Elements: " .. tostring(#self.Elements))
    print("Total Buttons: " .. tostring(#self.Buttons))
    print("====================================\n")
end

return ModMenu

File: modmenu/modmenu.js
// ModMenu UI JavaScript Implementation

class ModMenuUI {
    constructor(config = {}) {
        this.config = config;
        this.isOpen = false;
        this.elements = {};
        this.pages = {};
        this.currentPage = config.defaultPage || 'main';
        this.init();
    }

    init() {
        this.createHTML();
        this.attachEventListeners();
        console.log('[ModMenuUI] Initialized');
    }

    createHTML() {
        const container = document.createElement('div');
        container.className = 'modmenu-container';
        container.id = 'modmenu';
        container.innerHTML = `
            <div class="modmenu-header">
                <h1>${this.config.title || 'MOD MENU'}</h1>
                <button class="modmenu-close-btn" id="modmenuClose">✕</button>
            </div>
            <div class="modmenu-pages" id="modmenuPages"></div>
            <div class="modmenu-content" id="modmenuContent"></div>
            <div class="modmenu-footer">
                ${this.config.title || 'MOD MENU'} v1.0.0
            </div>
        `;
        document.body.appendChild(container);
    }

    attachEventListeners() {
        document.getElementById('modmenuClose').addEventListener('click', () => this.toggle());
        
        // Hotkey: M to toggle menu
        document.addEventListener('keydown', (e) => {
            if (e.key.toLowerCase() === 'm' && e.altKey) {
                this.toggle();
            }
        });
    }

    toggle() {
        this.isOpen = !this.isOpen;
        const container = document.getElementById('modmenu');
        container.classList.toggle('open', this.isOpen);
        console.log('[ModMenuUI] Menu is now ' + (this.isOpen ? 'OPEN' : 'CLOSED'));
    }

    addPage(pageId, pageTitle) {
        this.pages[pageId] = { id: pageId, title: pageTitle, elements: [] };
        this.renderPageTabs();
    }

    renderPageTabs() {
        const pagesContainer = document.getElementById('modmenuPages');
        pagesContainer.innerHTML = '';
        
        for (const [pageId, page] of Object.entries(this.pages)) {
            const btn = document.createElement('button');
            btn.className = `modmenu-page-btn ${pageId === this.currentPage ? 'active' : ''}`;
            btn.textContent = page.title;
            btn.addEventListener('click', () => this.switchPage(pageId));
            pagesContainer.appendChild(btn);
        }
    }

    switchPage(pageId) {
        if (this.pages[pageId]) {
            this.currentPage = pageId;
            this.renderPageTabs();
            this.renderPageContent();
        }
    }

    addButton(elementId, label, callback, config = {}) {
        this.addElement(elementId, {
            type: 'button',
            label: label,
            callback: callback,
            color: config.color || '#4CAF50',
            page: config.page || this.currentPage
        });
    }

    addToggle(elementId, label, defaultState, callback, config = {}) {
        this.addElement(elementId, {
            type: 'toggle',
            label: label,
            state: defaultState,
            callback: callback,
            page: config.page || this.currentPage
        });
    }

    addSlider(elementId, label, min, max, defaultValue, callback, config = {}) {
        this.addElement(elementId, {
            type: 'slider',
            label: label,
            min: min,
            max: max,
            value: defaultValue,
            callback: callback,
            page: config.page || this.currentPage
        });
    }

    addDropdown(elementId, label, options, defaultValue, callback, config = {}) {
        this.addElement(elementId, {
            type: 'dropdown',
            label: label,
            options: options,
            selected: defaultValue,
            callback: callback,
            page: config.page || this.currentPage
        });
    }

    addTextInput(elementId, label, placeholder, callback, config = {}) {
        this.addElement(elementId, {
            type: 'input',
            label: label,
            placeholder: placeholder,
            callback: callback,
            page: config.page || this.currentPage
        });
    }

    addElement(elementId, config) {
        this.elements[elementId] = config;
        const pageId = config.page || this.currentPage;
        if (this.pages[pageId] && !this.pages[pageId].elements.includes(elementId)) {
            this.pages[pageId].elements.push(elementId);
        }
        if (pageId === this.currentPage) {
            this.renderPageContent();
        }
    }

    renderPageContent() {
        const contentDiv = document.getElementById('modmenuContent');
        contentDiv.innerHTML = '';
        
        const page = this.pages[this.currentPage];
        if (!page) return;

        page.elements.forEach(elementId => {
            const element = this.elements[elementId];
            if (!element) return;

            const elementDiv = document.createElement('div');
            elementDiv.className = 'modmenu-element';
            elementDiv.id = `elem-${elementId}`;

            switch (element.type) {
                case 'button':
                    elementDiv.innerHTML = `
                        <button class="modmenu-btn modmenu-btn-primary" id="btn-${elementId}">
                            ${element.label}
                        </button>
                    `;
                    document.getElementById(`btn-${elementId}`).addEventListener('click', element.callback);
                    break;

                case 'toggle':
                    elementDiv.innerHTML = `
                        <div class="modmenu-toggle">
                            <label class="modmenu-label">${element.label}</label>
                            <div class="modmenu-switch ${element.state ? 'on' : ''}" id="tog-${elementId}"></div>
                        </div>
                    `;
                    document.getElementById(`tog-${elementId}`).addEventListener('click', () => {
                        element.state = !element.state;
                        document.getElementById(`tog-${elementId}`).classList.toggle('on');
                        if (element.callback) element.callback(element.state);
                    });
                    break;

                case 'slider':
                    elementDiv.innerHTML = `
                        <label class="modmenu-label">${element.label}</label>
                        <div style="display: flex; align-items: center;">
                            <input type="range" class="modmenu-slider" id="sld-${elementId}" 
                                   min="${element.min}" max="${element.max}" value="${element.value}" />
                            <span class="modmenu-slider-value" id="sldval-${elementId}">${element.value}</span>
                        </div>
                    `;
                    const slider = document.getElementById(`sld-${elementId}`);
                    slider.addEventListener('input', (e) => {
                        element.value = e.target.value;
                        document.getElementById(`sldval-${elementId}`).textContent = e.target.value;
                        if (element.callback) element.callback(e.target.value);
                    });
                    break;

                case 'dropdown':
                    elementDiv.innerHTML = `
                        <label class="modmenu-label">${element.label}</label>
                        <div class="modmenu-dropdown" id="dd-${elementId}">
                            <button class="modmenu-dropdown-button" id="ddbtn-${elementId}">
                                <span>${element.selected}</span>
                                <span>▼</span>
                            </button>
                            <ul class="modmenu-dropdown-list" id="ddlist-${elementId}">
                                ${element.options.map(opt => `<li class="modmenu-dropdown-option ${opt === element.selected ? 'selected' : ''}" data-value="${opt}">${opt}</li>`).join('')}
                            </ul>
                        </div>
                    `;
                    const ddBtn = document.getElementById(`ddBtn-${elementId}`);
                    const ddList = document.getElementById(`ddlist-${elementId}`);
                    ddBtn.addEventListener('click', () => {
                        ddList.classList.toggle('open');
                        ddBtn.classList.toggle('open');
                    });
                    document.querySelectorAll(`#ddlist-${elementId} .modmenu-dropdown-option`).forEach(opt => {
                        opt.addEventListener('click', () => {
                            element.selected = opt.dataset.value;
                            document.getElementById(`ddBtn-${elementId}`).querySelector('span').textContent = opt.dataset.value;
                            ddList.querySelectorAll('.selected').forEach(s => s.classList.remove('selected'));
                            opt.classList.add('selected');
                            ddList.classList.remove('open');
                            ddBtn.classList.remove('open');
                            if (element.callback) element.callback(element.selected);
                        });
                    });
                    break;

                case 'input':
                    elementDiv.innerHTML = `
                        <label class="modmenu-label">${element.label}</label>
                        <input type="text" class="modmenu-input" id="inp-${elementId}" 
                               placeholder="${element.placeholder}" />
                    `;
                    document.getElementById(`inp-${elementId}`).addEventListener('input', (e) => {
                        element.value = e.target.value;
                        if (element.callback) element.callback(e.target.value);
                    });
                    break;
            }
            contentDiv.appendChild(elementDiv);
        });
    }

    showNotification(title, message, duration = 3, type = 'success') {
        const notif = document.createElement('div');
        notif.className = `modmenu-notification ${type}`;
        notif.innerHTML = `<strong>${title}</strong><br>${message}`;
        document.body.appendChild(notif);
        
        setTimeout(() => {
            notif.remove();
        }, duration * 1000);
    }

    getElementValue(elementId) {
        const element = this.elements[elementId];
        if (!element) return null;
        return element.value || element.state || element.selected || null;
    }

    setElementValue(elementId, value) {
        const element = this.elements[elementId];
        if (!element) return;
        
        if (element.type === 'toggle') {
            element.state = value;
        } else if (element.type === 'slider') {
            element.value = value;
        } else if (element.type === 'dropdown') {
            element.selected = value;
        } else if (element.type === 'input') {
            element.value = value;
        }
        this.renderPageContent();
    }
}

// Export for use
window.ModMenuUI = ModMenuUI;